x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/optimizer/rules/logical/preoptimizer/FoldInferenceFunctions.java

```diff
@@ -0,0 +1,131 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License
+ * 2.0; you may not use this file except in compliance with the Elastic License
+ * 2.0.
+ */
+
+package org.elasticsearch.xpack.esql.optimizer.rules.logical.preoptimizer;
+
+import org.elasticsearch.action.ActionListener;
+import org.elasticsearch.action.support.CountDownActionListener;
+import org.elasticsearch.xpack.esql.core.expression.Expression;
+import org.elasticsearch.xpack.esql.expression.function.inference.InferenceFunction;
+import org.elasticsearch.xpack.esql.inference.InferenceFunctionEvaluator;
+import org.elasticsearch.xpack.esql.optimizer.LogicalPreOptimizerContext;
+import org.elasticsearch.xpack.esql.plan.logical.LogicalPlan;
+
+import java.util.ArrayList;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+
+/**
+ * Pre-optimizer rule that performs constant folding for inference functions in logical plans.
+ * <p>
+ * This rule identifies inference functions with constant parameters and evaluates them at optimization time,
+ * replacing them with their computed results.
+ * <p>
+ * The folding process is recursive and handles nested inference functions by processing them in multiple
+ * passes until no more foldable functions remain.
+ * <p>
+ * Example transformation:
+ * {@code TEXT_EMBEDDING("hello world", "model1")} → {@code [0.1, 0.2, 0.3, ...]}
+ */
+public class FoldInferenceFunctions implements LogicalPlanPreOptimizerRule {
```

Comment:
I guess something like this would also be used for folding COMPLETION when it's used with a foldable prompt?

Reply:
You are right. There will be something similar for inference plans.